File: src/footer/menu.ts
/**
 * the reference design's footer-resident menu band.
 *
 * Menus render between the composer and the hint row, the reference design's inline-picker
 * shape (`picker_presentation.zig` / `catalog_screen_layout.zig` row format):
 *
 *   ── divider ──────────────
 *   Header 7 · Type to filter          3–9
 *   (blank)
 *     item        description
 *     item        description          ← selected row bold-white, no caret
 *   ── divider ──────────────
 *
 * The hint row itself is the statusline row; the app swaps its content while a
 * menu is open. Selection marker convention is the reference design's: none — bold vs dim.
 */

import { truncateToWidth, visibleWidth, type Component } from "@earendil-works/pi-tui";
import { theme } from "../render/style.ts";

export interface MenuItem {
	label: string;
	description?: string;
	/** Right-aligned dim metadata (the reference design's `resume-catalog · 8m · 12 turns`). */
	meta?: string;
	value: string;
}

export interface MenuSpec {
	title: string;
	items: MenuItem[];
	hint: string;
	/** Filter items as the user types after the trigger. */
	filter?: (query: string, items: MenuItem[]) => MenuItem[];
	onSelect: (item: MenuItem) => void;
	onClose: () => void;
}

const MAX_VISIBLE = 8;

export class FooterMenu implements Component {
	readonly spec: MenuSpec;
	query = "";
	selected = 0;
	private windowStart = 0;

	constructor(spec: MenuSpec) {
		this.spec = spec;
	}

	visibleItems(): MenuItem[] {
		const all = this.spec.items;
		if (this.query === "") return all;
		if (this.spec.filter) return this.spec.filter(this.query, all);
		const q = this.query.toLowerCase();
		return all.filter(
			(i) => i.label.toLowerCase().includes(q) || i.description?.toLowerCase().includes(q),
		);
	}

	move(delta: number): void {
		const n = this.visibleItems().length;
		if (n === 0) return;
		this.selected = (this.selected + delta + n) % n;
		if (this.selected < this.windowStart) this.windowStart = this.selected;
		if (this.selected >= this.windowStart + MAX_VISIBLE)
			this.windowStart = this.selected - MAX_VISIBLE + 1;
	}

	current(): MenuItem | undefined {
		return this.visibleItems()[this.selected];
	}

	setQuery(query: string): void {
		this.query = query;
		this.selected = 0;
		this.windowStart = 0;
	}

	invalidate(): void {}

	render(width: number): string[] {
		const t = theme();
		const items = this.visibleItems();
		const divider = t.fg("dim", "─".repeat(width));

		const count = items.length;
		const windowEnd = Math.min(this.windowStart + MAX_VISIBLE, count);
		const range =
			count > MAX_VISIBLE ? `${this.windowStart + 1}–${windowEnd}` : "";
		let header =
			t.bold(t.fg("userMessageText", this.spec.title)) +
			t.fg("muted", ` ${count}` + (this.query ? ` · ${this.query}` : " · Type to filter"));
		if (range) {
			const bare = header.replace(/\x1b\[[0-9;]*m/g, "");
			const pad = width - visibleWidth(bare) - range.length;
			if (pad > 1) header += " ".repeat(pad) + t.fg("muted", range);
		}

		const rows: string[] = [divider, header, ""];
		if (count === 0) {
			rows.push(t.fg("muted", "  Nothing found."));
		}
		// Column where descriptions start: after the longest visible label.
		const labelWidth = Math.min(
			36,
			Math.max(...items.slice(this.windowStart, windowEnd).map((i) => visibleWidth(i.label)), 0),
		);
		for (let i = this.windowStart; i < windowEnd; i++) {
			const item = items[i]!;
			const selected = i === this.selected;
			const label = item.label + " ".repeat(Math.max(0, labelWidth - visibleWidth(item.label)));
			let row = selected ? t.bold(t.fg("userMessageText", label)) : label;
			if (item.description) row += t.fg("muted", `  ${item.description}`);
			if (item.meta) {
				const bare = row.replace(/\x1b\[[0-9;]*m/g, "");
				const pad = width - 2 - visibleWidth(bare) - visibleWidth(item.meta);
				if (pad > 1) row += " ".repeat(pad) + t.fg("muted", item.meta);
			}
			rows.push(truncateToWidth(`  ${row}`, width, "…"));
		}
		rows.push(divider);
		return rows;
	}
}

File: src/tui/menu.rs
//! The footer picker band.
//!
//! Menus render between the composer and the status row, the reference
//! inline-picker shape:
//!
//! ```text
//!   ── divider ─────────────────────────────
//!   Commands 7 · Type to filter          1–7
//!
//!     /login   sign in to a provider
//!     /model   list or switch models       ← selected row bold, no caret
//!   ── divider ─────────────────────────────
//!   ↑↓ Navigate     Enter Use     Esc Close   (rides the status row)
//! ```
//!
//! Selection is bold-ink vs default — no caret glyph, the reference
//! convention.

use crate::tui::markdown::visible_width;
use crate::tui::render::bold;
use crate::tui::theme::Theme;

#[derive(Clone)]
pub struct MenuItem {
    pub label: String,
    pub description: String,
    /// Right-aligned dim metadata.
    pub meta: String,
    pub value: String,
}

impl MenuItem {
    pub fn new(label: &str, description: &str, value: &str) -> Self {
        MenuItem {
            label: label.into(),
            description: description.into(),
            meta: String::new(),
            value: value.into(),
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum MenuKind {
    Commands,
    Files,
    Models,
    Sessions,
    Skills,
}

pub struct Menu {
    pub kind: MenuKind,
    pub title: &'static str,
    pub hint: &'static str,
    items: Vec<MenuItem>,
    filtered: Vec<usize>,
    pub selected: usize,
    window_start: usize,
    query: String,
}

pub const HINT_USE: &str = "↑↓ Navigate     Enter Use     Esc Close";
/// The reference keeps six selectable rows below the header.
const MAX_VISIBLE: usize = 6;

/// Subsequence fuzzy match; lower score is better, None is no match.
pub fn fuzzy_score(query: &str, candidate: &str) -> Option<usize> {
    if query.is_empty() {
        return Some(usize::MAX / 2); // neutral: keep original order
    }
    let candidate_lower = candidate.to_lowercase();
    let query_lower = query.to_lowercase();
    let mut score = 0usize;
    let mut position = 0usize;
    let mut last_hit: Option<usize> = None;
    for needle in query_lower.chars() {
        let found = candidate_lower[position..].find(needle)?;
        let at = position + found;
        // Contiguity bonus: gaps cost, adjacency is free.
        if let Some(last) = last_hit {
            score += at - last - 1;
        } else {
            score += at; // earlier first hits rank higher
        }
        last_hit = Some(at);
        position = at + needle.len_utf8();
    }
    Some(score)
}

impl Menu {
    pub fn new(
        kind: MenuKind,
        title: &'static str,
        hint: &'static str,
        items: Vec<MenuItem>,
    ) -> Self {
        let mut menu = Menu {
            kind,
            title,
            hint,
            items,
            filtered: Vec::new(),
            selected: 0,
            window_start: 0,
            query: String::new(),
        };
        menu.refilter();
        menu
    }

    pub fn set_query(&mut self, query: &str) {
        if self.query != query {
            self.query = query.to_string();
            self.refilter();
        }
    }

    fn refilter(&mut self) {
        let mut scored: Vec<(usize, usize)> = self
            .items
            .iter()
            .enumerate()
            .filter_map(|(i, item)| {
                let best = fuzzy_score(&self.query, &item.label)
                    .into_iter()
                    .chain(fuzzy_score(&self.query, &item.description))
                    .min()?;
                Some((best, i))
            })
            .collect();
        scored.sort_by_key(|(score, i)| (*score, *i));
        self.filtered = scored.into_iter().map(|(_, i)| i).collect();
        self.selected = 0;
        self.window_start = 0;
    }

    pub fn len(&self) -> usize {
        self.filtered.len()
    }

    pub fn is_empty(&self) -> bool {
        self.filtered.is_empty()
    }

    pub fn current(&self) -> Option<&MenuItem> {
        self.filtered.get(self.selected).map(|&i| &self.items[i])
    }

    pub fn step(&mut self, delta: isize) {
        let n = self.filtered.len();
        if n == 0 {
            return;
        }
        self.selected = (self.selected as isize + delta).rem_euclid(n as isize) as usize;
        if self.selected < self.window_start {
            self.window_start = self.selected;
        }
        if self.selected >= self.window_start + MAX_VISIBLE {
            self.window_start = self.selected - MAX_VISIBLE + 1;
        }
    }

    pub fn render(&self, theme: &Theme, width: usize) -> Vec<String> {
        let count = self.filtered.len();
        let window_end = (self.window_start + MAX_VISIBLE).min(count);

        let mut header = format!(
            "{}{}",
            bold(&theme.fg("userMessageText", self.title)),
            theme.fg(
                "muted",
                &if self.query.is_empty() {
                    format!(" {count} · Type to filter")
                } else {
                    format!(" {count} · {}", self.query)
                }
            ),
        );
        if count > MAX_VISIBLE {
            let range = format!("{}–{}", self.window_start + 1, window_end);
            let pad = width.saturating_sub(visible_width(&header) + range.chars().count());
            if pad > 1 {
                header.push_str(&" ".repeat(pad));
                header.push_str(&theme.fg("muted", &range));
            }
        }

        let mut body: Vec<String> = Vec::new();
        if count == 0 {
            body.push(theme.fg("muted", "  Nothing found."));
        }
        // Reference treatment: selection is brightness alone — the selected
        // row's label is bold bright ink, every other row is dim entirely;
        // descriptions are always dim; three-space column gap.
        let label_width = self.filtered[self.window_start..window_end]
            .iter()
            .map(|&i| visible_width(&self.items[i].label))
            .max()
            .unwrap_or(0)
            .min(36);
        for slot in self.window_start..window_end {
            let item = &self.items[self.filtered[slot]];
            let selected = slot == self.selected;
            let padded = format!(
                "{}{}",
                item.label,
                " ".repeat(label_width.saturating_sub(visible_width(&item.label)))
            );
            let mut row = if selected {
                bold(&theme.fg("userMessageText", &padded))
            } else {
                theme.fg("dim", &padded)
            };
            if !item.description.is_empty() {
                row.push_str(&theme.fg("dim", &format!("   {}", item.description)));
            }
            if !item.meta.is_empty() {
                let pad = width.saturating_sub(2 + visible_width(&row) + visible_width(&item.meta));
                if pad > 3 {
                    row.push_str(&" ".repeat(pad));
                    row.push_str(&theme.fg("dim", &item.meta));
                }
            }
            let mut line = format!("  {row}");
            if visible_width(&line) > width {
                line = line
                    .chars()
                    .take(width.saturating_sub(1))
                    .collect::<String>()
                    + "…";
            }
            body.push(line);
        }
        crate::tui::panel::frame(theme, width, header, body)
    }
}

#[cfg(test)]
mod tests {
    use super::fuzzy_score;

    #[test]
    fn fuzzy_prefers_tight_early_matches() {
        // Subsequence required.
        assert!(fuzzy_score("xyz", "composer.rs").is_none());
        // Tighter match beats scattered.
        let tight = fuzzy_score("comp", "src/tui/composer.rs").unwrap();
        let scattered = fuzzy_score("comp", "core/completions.rs").unwrap();
        assert!(tight < usize::MAX / 2 && scattered < usize::MAX / 2);
        // Case-insensitive.
        assert!(fuzzy_score("LOG", "/login").is_some());
        // Empty query keeps everything, original order.
        assert_eq!(fuzzy_score("", "anything"), Some(usize::MAX / 2));
    }
}
